### PR TITLE
Schema mapping

### PR DESCRIPTION
**Pull request by sfc-gh-zefan**
Added schema mapping logic. Data will be loaded into respective column through name matching of the field.
Added logic to add the 'RECORD_CONTENT' column if not present.
Added a config to enable the above features.
Added tests.

### CODE REVIEW

**Comment by sfc-gh-japatel** [comment]
This PR has 330 commits, can you rebase master and force push?

File: src/main/java/com/snowflake/kafka/connector/SnowflakeSinkConnectorConfig.java [hunk 1]
```diff
@@ -152,6 +152,8 @@ public class SnowflakeSinkConnectorConfig {
           + "By default messages are not sent to the dead letter queue. "
           + "Requires property `errors.tolerance=all`.";
 
+  public static final String SCHEMATIZATION_ENABLE_CONFIG = "schematization.enable";
```
Comment (sfc-gh-japatel) [inline src/main/java/com/snowflake/kafka/connector/SnowflakeSinkConnectorConfig.java:155]: I would keep this as `enable.schematization`

File: src/main/java/com/snowflake/kafka/connector/Utils.java [hunk 2]
```diff
@@ -364,6 +364,16 @@ static String validateConfig(Map<String, String> config) {
       if (!BufferThreshold.validateBufferThreshold(config, IngestionMethodConfig.SNOWPIPE)) {
         configIsValid = false;
       }
+      if (config.containsKey(SnowflakeSinkConnectorConfig.SCHEMATIZATION_ENABLE_CONFIG)) {
```
Comment (sfc-gh-japatel) [inline src/main/java/com/snowflake/kafka/connector/Utils.java:367]: did you mean to check `snowpipe_streaming` here?

File: src/main/java/com/snowflake/kafka/connector/Utils.java [hunk 3]
```diff
@@ -364,6 +364,16 @@ static String validateConfig(Map<String, String> config) {
...
+        boolean enableSchematization =
+            Boolean.parseBoolean(
+                config.get(SnowflakeSinkConnectorConfig.SCHEMATIZATION_ENABLE_CONFIG));
+        if (enableSchematization) {
+          configIsValid = false;
+          LOGGER.error(
+              Logging.logMessage("Schematization is only available with SNOWPIPE_STREAMING."));
```
Comment (sfc-gh-japatel) [inline src/main/java/com/snowflake/kafka/connector/Utils.java:374]: Lets use the enum here for printing.. `IngestionMethodConfig`

File: src/main/java/com/snowflake/kafka/connector/internal/SnowflakeConnectionService.java [hunk 4]
```diff
@@ -234,4 +235,12 @@ public interface SnowflakeConnectionService {
 
   /** @return the raw jdbc connection */
   Connection getConnection();
+
+  /**
+   * append a VARIANT type column "RECORD_METADATA" to the table if not present only called when
```
Comment (sfc-gh-japatel) [inline src/main/java/com/snowflake/kafka/connector/internal/SnowflakeConnectionService.java:240]:
```
<p> Append a VARIANT type column "RECORD_METADATA" to the table if it is not present.

<p> This method is only called when schematization is enabled
```

File: src/main/java/com/snowflake/kafka/connector/internal/SnowflakeConnectionService.java [hunk 5]
```diff
@@ -234,4 +235,12 @@ public interface SnowflakeConnectionService {
...
+   * schematization enabled
+   *
+   * @param tableName table name
+   */
```
Comment (sfc-gh-japatel) [inline src/main/java/com/snowflake/kafka/connector/internal/SnowflakeConnectionService.java:244]:
```suggestion

  /**
   * <p> Append a VARIANT type column "RECORD_METADATA" to the table if it is not present.
   * <p> This method is only called when schematization is enabled
   *
   * @param tableName table name
   */
```

File: src/main/java/com/snowflake/kafka/connector/Utils.java (hunk 2 above)
Comment (sfc-gh-zefan) [reply src/main/java/com/snowflake/kafka/connector/Utils.java:367]: This check is to make sure snowpipe & enable.schematization are not set at the same time.

File: src/main/java/com/snowflake/kafka/connector/SnowflakeSinkConnectorConfig.java (hunk 1 above)
Comment (sfc-gh-zefan) [reply src/main/java/com/snowflake/kafka/connector/SnowflakeSinkConnectorConfig.java:155]: Done.

File: src/main/java/com/snowflake/kafka/connector/Utils.java (hunk 2 above)
Comment (sfc-gh-japatel) [reply src/main/java/com/snowflake/kafka/connector/Utils.java:367]: nvm, didnt see the above if case. Thanks

File: src/main/java/com/snowflake/kafka/connector/internal/SnowflakeConnectionServiceV1.java [hunk 6]
```diff
@@ -326,6 +326,49 @@ public boolean isTableCompatible(final String tableName) {
...
+    boolean isVariant = false;
+    try {
+      stmt = conn.prepareStatement(query);
+      stmt.setString(1, tableName);
+      result = stmt.executeQuery();
+      while (result.next()) {
+        if (result.getString(1).equals("RECORD_METADATA")) {
+          hasMeta = true;
+          if (result.getString(2).equals("VARIANT")) {
+            isVariant = true;
```
Comment (sfc-gh-japatel) [inline src/main/java/com/snowflake/kafka/connector/internal/SnowflakeConnectionServiceV1.java:346]: nit: These are all magic numbers here. (1) and (2). 

Might be good to explain what 1 and 2 means here. column name and data type but would be a good call out..

File: src/main/java/com/snowflake/kafka/connector/internal/SnowflakeConnectionServiceV1.java [hunk 7]
```diff
@@ -326,6 +326,49 @@ public boolean isTableCompatible(final String tableName) {
...
+    boolean hasMeta = false;
+    boolean isVariant = false;
+    try {
+      stmt = conn.prepareStatement(query);
+      stmt.setString(1, tableName);
+      result = stmt.executeQuery();
+      while (result.next()) {
+        if (result.getString(1).equals("RECORD_METADATA")) {
+          hasMeta = true;
+          if (result.getString(2).equals("VARIANT")) {
```
Comment (sfc-gh-japatel) [inline src/main/java/com/snowflake/kafka/connector/internal/SnowflakeConnectionServiceV1.java:345]: will this always be variant or VARIANT? needs `equals ignorecase`?

File: src/main/java/com/snowflake/kafka/connector/internal/SnowflakeConnectionServiceV1.java [hunk 8]
```diff
@@ -326,6 +326,49 @@ public boolean isTableCompatible(final String tableName) {
...
+    PreparedStatement stmt = null;
+    ResultSet result = null;
+    boolean hasMeta = false;
+    boolean isVariant = false;
+    try {
+      stmt = conn.prepareStatement(query);
+      stmt.setString(1, tableName);
+      result = stmt.executeQuery();
+      while (result.next()) {
+        if (result.getString(1).equals("RECORD_METADATA")) {
```
Comment (sfc-gh-japatel) [inline src/main/java/com/snowflake/kafka/connector/internal/SnowflakeConnectionServiceV1.java:343]: consider extracting "RECORD_METADATA" to a static final string, something like
```
private static final String METADATA_COLUMN = "RECORD_METADATA"
```

and then using it here. I know table creation has not used it but may be we can start.

**Comment by sfc-gh-japatel** [comment]
Why closed? @sfc-gh-zefan

**Comment by sfc-gh-japatel** [comment]
Can you also add a JIRA?

**Comment by sfc-gh-zefan** [comment]
I tried to do the rebase and force push. This will automatically close the related pr.

**Comment by sfc-gh-zefan** [comment]
During rebasing there might be something wrong and a pr couldn't be created. I might create a brand new branch and copy all the changes so that it is clean. I will update the suggested changes on the way.

**Pull request closed without merging**